src/middleware/persist.ts

```diff
@@ -144,188 +124,207 @@ const toThenable =
     }
   }
 
-export const persist =
-  <
-    S extends State,
-    CustomSetState extends SetState<S> = SetState<S>,
-    CustomGetState extends GetState<S> = GetState<S>,
-    CustomStoreApi extends StoreApi<S> = StoreApi<S>
-  >(
-    config: (
-      set: CustomSetState,
-      get: CustomGetState,
-      api: CustomStoreApi
-    ) => S,
-    baseOptions: PersistOptions<S>
-  ) =>
-  (
-    set: CustomSetState,
-    get: CustomGetState,
-    api: CustomStoreApi & StoreApiWithPersist<S>
-  ): S => {
-    let options = {
-      getStorage: () => localStorage,
-      serialize: JSON.stringify as (state: StorageValue<S>) => string,
-      deserialize: JSON.parse as (str: string) => StorageValue<Partial<S>>,
-      partialize: (state: S) => state,
-      version: 0,
-      merge: (persistedState: any, currentState: S) => ({
-        ...currentState,
-        ...persistedState,
-      }),
-      ...baseOptions,
-    }
-
-    let hasHydrated = false
-    const hydrationListeners = new Set<PersistListener<S>>()
-    const finishHydrationListeners = new Set<PersistListener<S>>()
-    let storage: StateStorage | undefined
-
-    try {
-      storage = options.getStorage()
-    } catch (e) {
-      // prevent error if the storage is not defined (e.g. when server side rendering a page)
-    }
-
-    if (!storage) {
-      return config(
-        ((...args) => {
-          console.warn(
-            `[zustand persist middleware] Unable to update item '${options.name}', the given storage is currently unavailable.`
-          )
-          set(...args)
-        }) as CustomSetState,
-        get,
-        api
-      )
-    }
+const persistImpl: PersistImpl = (config, baseOptions) => (set, get, api) => {
+  type S = ReturnType<typeof config>
+  let options = {
+    getStorage: () => localStorage,
+    serialize: JSON.stringify as (state: StorageValue<S>) => string,
+    deserialize: JSON.parse as (str: string) => StorageValue<S>,
+    partialize: (state: S) => state,
+    version: 0,
+    merge: (persistedState: unknown, currentState: S) => ({
+      ...currentState,
+      ...(typeof persistedState === 'object' ? persistedState : {}),
```

Comment:
`persistedState` comes from the (local)storage so it can be anything (hence it's `unknown`), for eg  it could even be a number, technically `{ ...3 }` doesn't throw but typescript doesn't allow spreading non-objects hence the change.

Reply:
Yes it's kinda a TS issue. TS is conservative when it comes to spreading so that you don't accidentally spread a non-object (as most likely it'd be a mistake because it's not semantically correct). But for our case this acts like a false positive, so we can do `{ ...(persistedState as object) }` (or even `any` for that matter). Would you prefer that? Idm eitherway.

(I did the object check because I also consider spreading non-objects semantically wrong even if js doesn't throw)